lib/index.js

```diff
@@ -0,0 +1,146 @@
+'use strict'
+const Promise = require('bluebird')
+const yaml = require('js-yaml')
+const Joi = require('joi')
+const hostname = require('@runnable/hostname')
+
+const removeNullAndUndefinedKeys = require('./util').removeNullAndUndefinedKeys
+const validate = require('./util').validate
+const Parser = require('./parser')
+const Warning = require('./warning')
+
+const dockerComposeFileSchema = Joi.object({
+  version: Joi.string().valid('2').required(),
+  services: Joi.object().required()
+}).unknown().required()
+
+const arrayOfStrings = Joi.array().items(Joi.string())
+
+const serviceSchemaKeys = {
+  build: Joi.alternatives().try(Joi.string(), Joi.object()),
+  image: Joi.string(), // TODO: Add parser
+  command: Joi.alternatives().try(Joi.string(), arrayOfStrings),
+  ports: arrayOfStrings,
+  expose: arrayOfStrings, // TODO
+  links: arrayOfStrings, // TODO:
+  depends_on: arrayOfStrings, // TODO
+  environment: arrayOfStrings
+}
+const serviceSchema = Joi.object(serviceSchemaKeys).unknown().required()
+
+module.exports = class DockerComposeParser {
+
+  static parse (opts) {
+    const dockerComposeFileString = opts.dockerComposeFileString
+    const repositoryName = opts.repositoryName
+    const userContentDomain = opts.userContentDomain
+    const ownerUsername = opts.ownerUsername
+    return Promise.try(() => {
+      const dockerCompose = yaml.safeLoad(dockerComposeFileString)
+      return validate(dockerComposeFileSchema, dockerCompose)
+      .then(dockerCompose => dockerCompose.services)
+    })
+    .then(services => {
+      const mainServiceName = DockerComposeParser.getMainName(services)
```

Comment:
This isa stylisic thing, but what if you broke these out into different functions with descriptive names? Then  I just read it like a poem and then I read the specific functions and what they do. Totally optional just throwing it out there 😉  😜